Kuadrant/limitador-operator
# feat: resource requirements

# Description
* Use default and allow specifying resource requirements for limitador pod
* Use ginkgo v2 as v1 is deprecated
* Increase unit test coverage of `/api`

Closes: https://github.com/Kuadrant/limitador-operator/issues/74

# Verification
Functionality is generally tested by the included tests already but you can run through the following for a more thorough verification:

* Checkout this branch
* Deploy local cluster and build
```sh
make local-setup
```
* Deploy limitador cr with no resource requirements
```sh
kubectl apply -f config/samples/limitador_v1alpha1_limitador.yaml
```
* Verify default limits are set
```sh
kubectl get deployments limitador-sample -o json | jq '.spec.template.spec.containers[0].resources'
```
![image](https://github.com/Kuadrant/limitador-operator/assets/24636860/faab8ad4-f92c-432e-85ab-681a835be5a5)
* Patch Limitatdor CR with resource limits
```sh
kubectl apply -f -<<EOF                                          
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador-sample
spec:
  listener:
    http:
      port: 8080
    grpc:
      port: 8081
  limits:
    - conditions: ["get_toy == 'yes'"]
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
  resourceRequirements:
    limits:
      cpu: 200m
      memory: 400Mi
    requests:
      cpu: 101m
      memory: 201Mi         
EOF
```
* Verify resource limits specified are set
```sh
kubectl get deployments limitador-sample -o json | jq '.spec.template.spec.containers[0].resources'
```
![image](https://github.com/Kuadrant/limitador-operator/assets/24636860/eddfca45-25ad-43a8-a8d1-a182bac50a8c)
* Patch CR with empty resource requirements (i.e. no limits)
```sh
kubectl apply -f -<<EOF                                                                           
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador         
metadata:               
  name: limitador-sample
spec:      
  listener:     
    http:       
      port: 8080
    grpc:       
      port: 8081                      
  limits:                             
    - conditions: ["get_toy == 'yes'"]
      max_value: 2           
      namespace: toystore-app
      seconds: 30      
      variables: []    
  resourceRequirements: {}
EOF
```
* Verify resource limits are empty
```sh
kubectl get deployments limitador-sample -o json | jq '.spec.template.spec.containers[0].resources'
```
![image](https://github.com/Kuadrant/limitador-operator/assets/24636860/5d613a06-70f2-4118-a676-a2bb6883f9ee)
* Patch CR to set resource requirements back to `nil`
```sh
kubectl apply -f -<<EOF                                                                           
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador         
metadata:               
  name: limitador-sample
spec:      
  listener:     
    http:       
      port: 8080
    grpc:       
      port: 8081                      
  limits:                             
    - conditions: ["get_toy == 'yes'"]
      max_value: 2           
      namespace: toystore-app
      seconds: 30      
      variables: []    
EOF
```
* Verifiy default limits are applied 
```sh
kubectl get deployments limitador-sample -o json | jq '.spec.template.spec.containers[0].resources'
```
![image](https://github.com/Kuadrant/limitador-operator/assets/24636860/cd49185a-c4d1-486a-b252-224cc7677a74)
